src/IO/H5/Dat.hpp

```diff
@@ -0,0 +1,110 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class h5::Dat
+
+#pragma once
+
+#include <array>
+#include <iosfwd>
+#include <vector>
+
+#include "IO/H5/Object.hpp"
+#include "IO/H5/OpenGroup.hpp"
+
+namespace h5 {
+/*!
+ * \ingroup HDF5
+ * \brief Represents a multicolumn dat file inside an HDF5 file
+ *
+ * Stores data that is normally stored inside a dat file but inside a dataset
+ * in an HDF5 file. This allows multiple Dat objects to be stored in a single
+ * HDF5 file, thus reducing the total number of files that need to be stored.
+ * This class does not do any caching of data so all data is written as soon
+ * as append() is called.
+ */
+class Dat : public h5::Object {
+ public:
+  /// \cond HIDDEN_SYMBOLS
+  static std::string extension() { return ".dat"; }
+
+  Dat(bool exists, detail::OpenGroup&& group, hid_t location,
+      const std::string& name, std::vector<std::string> legend = {},
+      uint32_t version = 1);
+
+  Dat(const Dat& /*rhs*/) = delete;
+  Dat& operator=(const Dat& /*rhs*/) = delete;
+  Dat(Dat&& /*rhs*/) noexcept = default;             // NOLINT
+  Dat& operator=(Dat&& /*rhs*/) noexcept = default;  // NOLINT
+
+  ~Dat() override;
+  /// \endcond HIDDEN_SYMBOLS
+
+  /*!
+   * \requires `data.size()` is the same as the number of columns in the file
+   * \effects appends `data` to the Dat file
+   */
+  void append(const std::vector<double>& data);
```

Comment:
I have not measured how expensive it is and I have zero intuition for HDF5 times. Based on what I expect HDF5 to be doing in the background writing multiple together would be noticeably faster, but I have nothing to back that claim up. 

As for the second question, I have no objection to adding an overload for both `std::vector<std::vector<...>>` and `std::array<std::vector<...>, #>`. They should be pretty easy to implement.